plutus-core/cost-model/test/TH.hs

```diff
@@ -0,0 +1,37 @@
+{- | Some basic Template Haskell to reduce boilerplate in the cost model tests.
+   We have to put this in a separate source file because of staging
+   restrictions.
+-}
+
+{-# LANGUAGE TemplateHaskell #-}
+
+module TH (genTest)
+where
+
+import Data.Char (toUpper)
+import Language.Haskell.TH
+
+toUpper1 :: String -> String
+toUpper1 []     = error "empty string in toUpper1"
+toUpper1 (c:cs) = (toUpper c):cs
+
+mkIterApp :: Exp -> [Exp] -> Exp
```

Comment:
there's https://hackage.haskell.org/package/template-haskell-2.13.0.0/docs/Language-Haskell-TH-Lib.html#v:appsE if you're willing to do the weird TH thing of operating with `ExprQ` everywhere instead of `Expr` (which I've always found odd...)